src/librustc/mir/interpret/cast.rs

```diff
@@ -93,17 +94,26 @@ impl<'a, 'tcx, M: Machine<'tcx>> EvalContext<'a, 'tcx, M> {
         }
     }
 
-    fn cast_from_float(&self, val: f64, ty: Ty<'tcx>) -> EvalResult<'tcx, PrimVal> {
+    fn cast_from_float(&self, val: ConstFloat, ty: Ty<'tcx>) -> EvalResult<'tcx, PrimVal> {
         use ty::TypeVariants::*;
         match ty.sty {
-            // Casting negative floats to unsigned integers yields zero.
-            TyUint(_) if val < 0.0 => self.cast_from_int(0, ty, false),
-            TyInt(_) if val < 0.0 => self.cast_from_int(val as i128 as u128, ty, true),
+            TyUint(t) => {
+                let width = t.bit_width().unwrap_or(self.memory.pointer_size() as usize * 8);
+                match val.ty {
+                    FloatTy::F32 => Ok(PrimVal::Bytes(Single::from_bits(val.bits).to_u128(width).value)),
```

Comment:
So this saturates. HIR const eval doesn't saturate, it reports an error instead. I don't think there's been a formal decision to do saturation, so it may be prudent to keep reporting the error. For context, #45205 keeps the error in HIR const eval, and I'll add the same error to trans const eval, at least to see if crater reports any regressions.